Source/Core/Core/State.cpp

```diff
@@ -835,87 +824,78 @@ static void LoadFileStateData(const std::string& filename, Common::UniqueBuffer<
   ret_data.swap(buffer);
 }
 
-void LoadAs(Core::System& system, const std::string& filename)
+static void LoadAsFromCore(Core::System& system, std::string filename)
 {
-  if (!Core::IsRunningOrStarting(system))
-    return;
-
-  if (NetPlay::IsNetPlayRunning())
+  // Save temp buffer for undo load state
+  auto& movie = system.GetMovie();
+  if (!movie.IsJustStartingRecordingInputFromSaveState())
   {
-    OSD::AddMessage("Loading savestates is disabled in Netplay to prevent desyncs");
-    return;
+    SaveToBuffer(system, s_undo_load_buffer);
+    const std::string dtmpath = File::GetUserPath(D_STATESAVES_IDX) + "undo.dtm";
+    if (movie.IsMovieActive())
+      movie.SaveRecording(dtmpath);
+    else if (File::Exists(dtmpath))
+      File::Delete(dtmpath);
   }
 
-  if (AchievementManager::GetInstance().IsHardcoreModeActive())
+  bool was_file_read = false;
+  bool loaded_successfully = false;
+
+  // brackets here are so buffer gets freed ASAP
   {
-    OSD::AddMessage("Loading savestates is disabled in RetroAchievements hardcore mode");
-    return;
-  }
+    WaitForStateSaveTaskCompletion();
 
-  std::unique_lock lk(s_load_or_save_in_progress_mutex, std::try_to_lock);
-  if (!lk)
-    return;
+    Common::UniqueBuffer<u8> buffer;
+    LoadFileStateData(filename, buffer);
```

Comment:
This one looks dangerous. You're calling `WaitForStateSaveTaskCompletion`, but after that there's nothing stopping another thread from saving a state while you're loading it.

Reply:
No, hold on... Both SaveAsFromCore and LoadAsFromCore are only allowed to run on the CPU thread, so they can't both run at once. I *think* this is safe then, but the novelty of this way of handling synchronization makes it harder to reason about. (Maybe the old code is just as hard to reason about, though? I haven't looked at it closely.)

Reply:
Yeah, it can still be a little hard to reason about, but I think it was a lot worse, there's four mutexes in master! `s_undo_load_buffer_mutex` `s_load_or_save_in_progress_mutex` `s_save_thread_mutex` `s_state_writes_in_queue_mutex`. I think it was infeasible to reason out whether or not they were going to deadlock. And all those mutexes and they were still producing save/load races.

Reply:
>If that were to be the case, LoadFileStateData should lock s_state_fs_mutex.

I'm guessing I did that so e.g. `LoadFileStateData` and `GetInfoStringOfSlot` wouldn't contest the mutex, but it might be easier to follow if I just made `s_state_fs_mutex` a `shared_mutex` and made all filesystem access lock it appropriately.

Reply:
Okay, doing exactly that was being messy so I've instead eliminated `s_state_fs_mutex` entirely.
I've replaced my `WaitableCounter` with a less-weird `TransferableSharedMutex` which is now the only mutex.
`GetInfoStringOfSlot` and friends now hold this mutex for the entirety of the call which also resolves the previously mentioned "benign race".
I've also added some comments about things that seemed unclear.